improver/expected_value.py

```diff
@@ -0,0 +1,76 @@
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Calculation of expected value from a probability distribution."""
+
+import iris.analysis
+from iris.coords import CellMethod
+from iris.cube import Cube
+
+from improver import PostProcessingPlugin
+from improver.ensemble_copula_coupling.ensemble_copula_coupling import (
+    ConvertProbabilitiesToPercentiles,
+    RebadgePercentilesAsRealizations,
+)
+from improver.metadata.probabilistic import is_percentile, is_probability
+from improver.utilities.cube_manipulation import collapsed
+
+
+class ExpectedValue(PostProcessingPlugin):
+    """Calculation of expected value from a probability distribution"""
+
+    def process(self, cube: Cube) -> Cube:
+        """Expected value calculation and metadata updates.
+
+        Args:
+            cube:
+                Probabilistic data with a realization, threshold or percentile
+                representation.
+
+        Returns:
+            Expected value of probability distribution. Same shape as input cube
+            but with realization/threshold/percentile coordinate removed.
+        """
+        if is_probability(cube):
+            # TODO: replace this with direct calculation of the integral over
+            # probability thresholds. The current approach works and has the
+            # correct interface, but some accuracy will be lost in the conversion
+            # and memory usage is very high.
+            #
+            # 19 percentiles corresponds to 5, 10, 15...95%.
+            cube = ConvertProbabilitiesToPercentiles().process(
+                cube, no_of_percentiles=19
+            )
+        if is_percentile(cube):
+            cube = RebadgePercentilesAsRealizations().process(cube)
+        mean_cube = collapsed(cube, "realization", iris.analysis.MEAN)
+        mean_cube.remove_coord("realization")
+        mean_cube.add_cell_method(CellMethod("mean", coords="realization"))
```

Comment:
The threshold, percentile and realization representations are all numerical approximations of the underlying probability density function (PDF) representing some weather parameter.
There are three cases to consider here.

For realization data:
 - There is an assumption that all realizations have equal probability, so the weighted mean definition of expected value reduces to the plain (unweighted) arithmetic mean of the realizations.
    + See the sentence above the "Examples" heading under [Random variables with finitely many outcomes](https://en.wikipedia.org/wiki/Expected_value#Random_variables_with_finitely_many_outcomes) which talks about this case.
- The realizations being considered to have equal probability is common in ensemble NWP model data. The control member may be slightly "better" and more likely to occur than other ensemble members, for example due to not having any perturbation applied to the initial conditions. However, I haven't seen any regularly used quantification of "how much more likely is the control member". My experience is that it is generally bundled in with the other ensemble members and treated equally.
- If reviewers think it would be useful to explain this in the docstrings, I can add that.

For percentile data:
- If the percentiles are equally spaced between 0% and 100%, there is equal probability of each percentile value occurring. This is the underpinning of RebadgePercentilesAsRealisations and a stepping stone towards ensemble copula coupling.
- Given equal probability of each percentile value occurring, the above points about realization data apply here too.
 
For threshold data:
- The expected value is an integral of the PDF, weighted by the x-axis values.
    + See the first equation under [Random variables with density](https://en.wikipedia.org/wiki/Expected_value#Random_variables_with_density).
- There's no need to provide separate weights – the weights are provided by the values of the threshold coordinate in the input cube.
- Threshold data is a numerical approximation of the CDF, so implementing this calculation would first convert from the CDF representation (eg. probability of temperature greater than 273 kelvin) to a PDF over discrete intervals (eg. probability of temperature between 273 and 274 kelvin), then do numerical integration (eg. midpoint rule or trapezoidal rule) over those intervals.
- This numerical integration process isn't implemented in this PR, see comment in the code.

The three cases are all mathematically equivalent (ie. given one definition, a proof could be written to derive the other definitions). I don't think there's any reason that one should be named differently to another or considered as the "true" expected value.

Reply:
Thanks for giving this serious consideration. I agree with your points above, namely that given the equal-probability assumption for ensemble members that the ensemble mean is mathematically equivalent to expectation value. I think you've sufficiently justified leaving it as is (particularly given the intention is to implement the numerical integration on threshold data in future).